### ISSUE TITLE
Cross-domain behavior of localstorage+cookie

### ISSUE THREAD

**Issue by dzsibi** [PostHog/posthog-js#1399]
Hello,

I am somewhat confused about how `localstorage+cookie` works across different subdomains. The current implementation merges the values stored in the cookie with those stored in local storage, preferring the values in local storage:

https://github.com/PostHog/posthog-js/blob/34204f4aff107162bdb259d2a7c855d67bf0260c/src/storage.ts#L262

But this creates a very undesirable behavior when the user navigates between different subdomains. Take the following sequence of events:

1. User visits `www.example.com`, is assigned a random distinct ID, which is stored in a cookie (for `.example.com`) and in local storage as well (for `www.example.com`)
2. User then navigates to `support.example.com`, where the distinct ID is loaded from the cookie (for `.example.com`) which is also saved to local storage there (for `support.example.com`).
3. The user logs in on `support.example.com`, identify is called, and now has an authenticated distinct ID, which overwrites the value in the cookie (for `.example.com`) and in local storage (for `support.example.com`)
4. The user navigates back to `www.example.com`, where the authenticated distinct ID is loaded from the cookie (for `.example.com`), but is overwritten by the anonymous distinct ID in local storage (for `www.example.com`). Afterwards, the anonymous distinct ID also overwrites the value in the cookie (for `.example.com`).
5. (If the user then visits a third subdomain, the authenticated distinct ID is already lost.)

In most use cases, I think that if the user is identified, that information should spread across subdomains. Of course one could keep using the cookie-only storage, which solves the issue, but makes all requests to the server larger and can cause problems if the cookie grows too large (which is why I gather `localstorage+cookie` was made the default persistence option).

One way to fix this would be to prefer the values in the cookie, effectively reversing the merge order. Additional logic may be added to avoid merging values between different authenticated users. A conservative approach would be:

- If the distinct ID in the cookie and local storage is the same, merge without further checks
- If only one of them is authenticated, prefer that value, and discard the contents of the other
- Otherwise (if they are for different, authenticated users), do not merge

A more aggressive approach would be that always prefers the value in the cookie:

- If the distinct ID in the cookie and local storage is the same, merge without further checks
- Otherwise, discard value in local storage and use the one from the cookie

Is the current behavior by design? Would it be possible to change it / make it configurable, or should we simply roll our own PersistentStore to work around it?

**Comment by ksmoore17** [comment]
I have a similar issue with the cross subdomain behavior. I have one site that performs auth and thus uses `reset` and `identify`. I have another site on a sibling subdomain (that has no auth capabilities) picking up the cross subdomain cookie and using its identification/posthog setup.
When I have both sites loaded on two tabs, and I trigger either reset or identify on the authed site, the cookie/localStorage updates with an expected distinct id (either generated by ph or my provided distinct id based on the user), but then reloading or navigating on the no-auth site sets the cookie/localStorage persisted data to what it was before the reset or identify.

#987 relates as even after manually deleting posthog data through the browser's inspector in cookies and localStorage, the posthog client on the no-auth site sets the cookie and localStorage entry on reload or navigation. This leads me to believe that the posthog data/distinct id stored in memory is used to override the persistence on event capture. In fact, I can see that not capturing `pageleave` events while still capturing `pageview` events causes the persistence to remain untouched on _reload_, but the cookie setting still happens on _navigate_. I can see that the [`capture`](https://github.com/PostHog/posthog-js/blob/909eda0cd1a9b295078c2fa6d86d4e9f7b18780e/src/posthog-core.ts#L799) method may call several other methods that call `persistence.save`.

Ultimately, I would like to effectively ignore the data that the ph client has in memory, using the cookie as the single source of truth (the behavior I expected), as I don't have a way to know when to reset on my no-auth site. Even if I did, I would be resetting twice as I still would have my authed site resetting first.

**Comment by rafaeelaudibert** [comment]
Hey, guys!

I believe there's some confusion on how `distinct_id` works. We recommend against setting `distinct_id` manually. You should always be setting `person_id`, and then once you call `identify(...)` we'll know that we need to link that random anon `person_id` to the `distinct_id` you passed in `identify`. This guarantees that it doesn't matter whether you're calling `capture` (or any other function) with the original anon ID or the new distinct ID those events will always end up under the same person inside PostHog.

Does that make sense?

**Comment by ksmoore17** [comment]
I don't think we are setting it manually, and the problem comes in part because one of the sites in question does not or cannot call `identify`.

This is what I am seeing and it think @dzsibi has noticed the same or similar:

1. Site A is loaded, posthog inits and creates an anonymous `distinct_id`
2. Site B is loaded on another tab, posthog inits and calls `identify` when the user logs in -> the cross subdomain cookie is set and the `localStorage` for that tab B contains the same `distinct_id` as the cookie
3. Site A, back on the same first tab, captures an event (like pageleave) without having to `init` posthog again (in the same tab as step 1, so `localStorage` AND the client in memory are unchanged from the site A `init`)

On that second "action" (the pageleave capture) on site A, site A has not called `identify` or `init`, and therefore has a mismatch between the cross subdomain cookie that Site B has set and the information that the posthog client has in memory. It captures an event with no regard to the identified `distinct_id` that has been set in the cookie.

I think that the difference between our two examples is that the original example has a full page load back to site A (`www.example.com`). This means that posthog `init` is called again on site A (and therefore `load`). Because of the "merge order" of `extend` called in `load`, the `distinct_id` from `localStorage` has "priority" over the `distinct_id` from the cookie set by site B (`support.example.com`), meaning the client is created with the `distinct_id` from site A's "out of date" `localStorage`.
In my example, there is no second `init`/`load` call on site A as you are just going back to the tab that already has the initialized client, but now has a different ph cookie. So the ph client just uses the `distinct_id` that it has in memory and even sets the cookie to the old anonymous id. So the `localStorage` mismatch is not relevant, but the cookie/memory mismatch is.

What I have just realized is that after identifying the anonymous `distinct_id` with a non-anonymous one, events submitted with the old anonymous distinct id will be still associated with the non-anonymous `distinct_id`, duh.

However, other things stored in both the cookie and `localStorage`, like whether the user has been identified (I think its the `$epp` property, right?), have the same "out of date" issue from not aggressively loading from the cookie (in my memory mismatch example), or prioritizing the cookie (the original `localStorage` mismatch example). This would lead to stuff like anonymous events being captured for users that had already been identified.

It is also relevant because when going through the same order but with `reset` instead of `identify` (like a logout on site A rather than a log in), site A would have events being submitted as the previously identified `distinct_id` from `localStorage`/memory, rather than with the new anonymous `distinct_id` from site B's `reset`.
